### ISSUE TITLE
[not an issue] Datasets used for training

### ISSUE THREAD

**Issue by nliaudat** [haverland/Tenth-of-step-of-a-meter-digit#25]
Hi, 

Many thanks for your great work ! 

I've read all your repo and I a bit confused with dataset used for training. 

I see references to 

- images/ziffer_raw
- images/collected (ok new from user)
- SVHN (Street View House Numbers) Dataset
- MNIST

Can you clarify the dataset used and their weight in training ? 

Perhaps could it be better to upload all the datasets to that repo (with augmented images) as there is no limits on github (except the max file size). 

Regards

**Comment by nliaudat** [comment]
many thanks !
I was using only https://github.com/haverland/Tenth-of-step-of-a-meter-digit/tree/master/datasets/meterdigits

**Comment by haverland** [comment]
> Ps. : this is part of my search of improvements for digit recognitions https://github.com/nliaudat/digit_recognizer/tree/main
> 
> It's very beta, the results do not fit real world, only training results.

Very interesting. Output ist 10 and I wonder why the best models only have an accuracy of 98%. Comparing to https://github.com/haverland/Tenth-of-step-of-a-meter-digit/blob/master/Compare_all_tflite.ipynb with +/-0.1 and 99,8% accuracy.

But the original tenth-step model is much oversized because it's designed for 100 outputs. 

Are you have tflites for testing?

**Comment by nliaudat** [comment]
Hi, 
This is still in early beta. 

The less accuracy can explain with many factors : 

- I was training on "kaggle" dataset with 18k images instead of 27k. 
- I apply "Quantization-Aware Training" on all layers
- The keras augmentation is not very clear in your notebook as there are references to augmentation/generator.py. 

I decided not to use notebooks and make tensorflow script from scratch for my own learnings.

**Comment by nliaudat** [comment]
Hi, 

I've introduced a bug with keras augmentation I have to fix. So the actual state in my graph is not relevant cause not using same dataset. 

The small accuracy of 10% you obtain is exactly what I have to fix in my side :  the augmentation data (or the dataset) is not passed in good format (float32, uint8 or int8)

Actually, the I am adding a final augmentation layer to normalize data : 

```
        augmentation_layers.append(
            tf.keras.layers.Lambda(
                lambda x: tf.clip_by_value(x, 0.0, 1.0),  # Ensure valid range
                name='final_value_clamp'
            )
        )
```

But It works only for QAT trained models. 

I have to work a bit more to get a all usage training function. 

Regards

**Comment by haverland** [comment]
Was incompatibility with TensorFlow and works now. I've removed the complete requirements and reinstalled it.

**Comment by nliaudat** [comment]
Hi, 
Your notebook performs float inference on quantized models without scaling or dequantization. 

This is why we have bad results. 

Can you add : 

```
def preprocess_input_for_model(image, input_details):
    """
    Prepares an input image for a TFLite model, applying quantization if needed.
    Handles uint8, int8, and float32 inputs.
    """
    input_dtype = input_details['dtype']
    input_scale, input_zero_point = input_details['quantization']

    # Convert to float32 for consistent normalization
    if image.dtype != np.float32:
        image = image.astype(np.float32)

    # Normalize to [0, 1] range if needed
    if image.max() > 1.0:
        image = image / 255.0

    # Apply quantization only for quantized models
    if input_dtype == np.uint8 or input_dtype == np.int8:
        # Quantized model expects integer inputs in quantized range
        image = image / input_scale + input_zero_point
        image = np.clip(image, np.iinfo(input_dtype).min, np.iinfo(input_dtype).max)
        image = image.astype(input_dtype)

    return image


def dequantize_output(output, output_details):
    """
    Converts a quantized model output back to float space.
    Handles both uint8 and int8 quantized outputs.
    """
    output_dtype = output_details['dtype']
    output_scale, output_zero_point = output_details['quantization']

    if output_dtype == np.uint8 or output_dtype == np.int8:
        output = output.astype(np.float32)
        output = (output - output_zero_point) * output_scale

    return output
```

and modify : 

`output = interpreter.get_tensor(output_index)`
to 
```
        output = interpreter.get_tensor(output_index)
        
        # Always dequantize output if needed
        output = dequantize_output(output, output_details)
```

and somethings like 


```
def evaluate_ziffer_tflite(model_path, x_data, x_data_gray, y_data, y_data10, f_data, title, max_delta = 0.11):
    false_images = []
    false_labels = []
    false_predicted = []

    # we use the tflite model
    print(model_path)
    interpreter = tf.lite.Interpreter(model_path=model_path)
    interpreter.allocate_tensors()
    input_details = interpreter.get_input_details()[0]
    output_details = interpreter.get_output_details()[0]
    input_index = input_details["index"]
    output_index = output_details["index"]

    print(f"Model input: shape={input_details['shape']}, dtype={input_details['dtype']}, quantization={input_details.get('quantization', 'None')}")
    print(f"Model output: shape={output_details['shape']}, dtype={output_details['dtype']}")

    # Preprocess all inputs based on model requirements
    processed_x_data = []
    for x in x_data:
        processed_x = preprocess_input_for_model(x, input_details)
        processed_x_data.append(processed_x)
    
    processed_x_data = np.array(processed_x_data)

    # Use appropriate labels based on output type
    output_shape = output_details['shape']
    if len(output_shape) == 1 or output_shape[-1] == 10:
        yd = y_data10  # Use integer labels for classification
    else:
        yd = y_data    # Use float labels for regression

    for x, y, f in zip(processed_x_data, yd, f_data):
        interpreter.set_tensor(input_index, np.expand_dims(x, axis=0))
        # Run inference.
        interpreter.invoke()
        # Post-processing: remove batch dimension and find the digit with highest
        # probability.
        output = interpreter.get_tensor(output_index)
        
        # Always dequantize output if needed
        output = dequantize_output(output, output_details)
        
        # FIXED PREDICTION LOGIC:
        output_vector = output[0]  # Remove batch dimension
        
        if len(output_vector) == 2:
            # Sin/Cos output type - outputs are [sin(2πx/10), cos(2πx/10)]
            out_sin = output_vector[0]  
            out_cos = output_vector[1]
            angle = np.arctan2(out_sin, out_cos)  # Returns angle in radians [-π, π]
            # Convert to digit 0-9
            prediction = (angle / (2 * math.pi)) % 1.0  # Normalize to [0, 1]
            prediction = prediction * 10.0  # Scale to [0, 10]
            if prediction >= 9.5:  # Handle wrap-around (9.9 -> 0.0)
                prediction = 0.0
                
        elif len(output_vector) == 10:
            # Classification output (10 classes) - apply softmax and get max
            # Apply softmax to get probabilities
            exp_output = np.exp(output_vector - np.max(output_vector))  # Numerical stability
            probabilities = exp_output / np.sum(exp_output)
            prediction = np.argmax(probabilities)
            confidence = np.max(probabilities)
            
        else:
            # Single output regression - output is already the predicted digit 0-9
            prediction = float(output_vector[0])
            # Ensure prediction is in valid range [0, 10)
            if prediction < 0:
                prediction = 0.0
            elif prediction >= 10:
                prediction = 0.0  # Wrap around

        # Calculate difference (handle circular nature of digits 0-9)
        difference = min(abs(prediction - y), 
                        abs(prediction - (y + 10)), 
                        abs(prediction - (y - 10)))
        
        if difference > max_delta:
            # Store original image for display, not processed one
            original_img_idx = np.where(f_data == f)[0][0]
            false_images.append(x_data[original_img_idx])  # Use original image
            false_labels.append(f"Expected: {y:.1f}\nPredicted: {prediction:.2f}\nFile: {os.path.basename(f)}")
            false_predicted.append(difference)

    # Rest of the function remains the same...
    total_flops = "{:.1f}  M FLOPS ".format( calc_flops(model_path) / 1e6)
    accuracy = "{:.2f}%".format((1-len(false_labels)/len(y_data))*100)
    title = f"Model: {os.path.basename(model_path)} ( {total_flops}), Tested images: {len(y_data)}. {len(false_labels)} false predicted. Accuracy is: {accuracy}\nDeviation of false predicted"
    
    if false_predicted:
        plot_divergence(np.bincount(np.array(np.array(false_predicted)*10).astype(int), minlength=51), title, 51)
        plot_dataset(np.array(false_images), false_labels, columns=7, rows=7, figsize=(18,18))
    else:
        print(f"Perfect accuracy! No false predictions.")
```

this is a not tested code, but there are some mismatches between our processing logic.

For the prediction, you apply softmax directly to quantized value and it can make a lot of bad values. 

As example, I get only 200 bad predicts on my models with 25000 images tested.

I have certainly made mistakes in my code cause it's my first shoot and many thanks for reporting the [4,7,5] bug.

My goal is only to improve the actual model, not to destroy what have been done for years.

**Comment by haverland** [comment]
The most false predictions have a difference of 1

<img width="1205" height="366" alt="Image" src="https://github.com/user-attachments/assets/4aa263e4-7e85-414b-81d6-0f04e9f0387d" />

And now look at the false predicted:

<img width="1207" height="377" alt="Image" src="https://github.com/user-attachments/assets/a169ef8d-1457-42ee-8dcc-0c3e36bb721a" />

In my comparison with class100 is a false prediction of +/- 0.1 and almost all not relevant. But for output=10 its much more difference. And the most false predicted you see are x.4,x.5,x.6.  

Since the other 91% were correctly recognised, the reason is not the transfer of the images, is it?

**Comment by nliaudat** [comment]
Hi, 

Very sorry, but I found that I have doubled quantization in my training. 
I rewrite all code and split into classes. 
The new models should provide correct predictions now (my prediction script was linked to training. Now I bench_predict.py separatly)

I've delete 5000 redundant images (1500 has same checksum) from your dataset cause they where overfitting  the model (https://github.com/nliaudat/digit_recognizer/blob/main/datasets/tools/deduplicate.py)

I mean you can test the digit_recognizer_v4.tflite safely. 

Regards

**Comment by haverland** [comment]
I will check removing the duplicates the next days. It makes sense.
